### PR TITLE
✨ Add Solution Explorer

### PR DESCRIPTION
**Pull request by Vyperus**
## What did you change?

This branch adds the SolutionExplorer to BPMN-Studio!

Waits on:

- [x] https://github.com/process-engine/solutionexplorer.service/pull/2
- [x] https://github.com/process-engine/bpmn-studio/pull/523

Relates to:

https://github.com/process-engine-pro/organization/issues/153

## How can others test the changes?

- Open BPMN-Studio with `npm run electron-start-dev`.
- Open SolutionExplorer.
- Click on Folder to see the FileSystem index tab.
- Click on "Open a Solution".
- Choose a folder with .bpmn files in it.

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.)
- [x] I've merged the `develop` branch into my branch before finishing this PR.
- [x] I've **not added any other changes** than the ones described above.
- [x] I've mentioned all **PRs, which relate to this one**

### CODE REVIEW

File: src/modules/diagram-detail/diagram-detail.ts [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+import {ICanvgOptions, NotificationType} from '../../contracts/index';
+import environment from '../../environment';
+import {BpmnIo} from '../bpmn-io/bpmn-io';
+import {NotificationService} from '../notification/notification.service';
+
+interface RouteParameters {
+  diagramName: string;
+}
+
+@inject('SolutionExplorerServiceFileSystem', 'NotificationService', EventAggregator, Router)
```
[the remark this answers is not in the document]
Comment (Vyperus) [reply src/modules/diagram-detail/diagram-detail.ts]: Ich finde es durchaus in Ordnung mehrere "views" zu haben deren Komponenten wissen wie sie mit der jeweiligen Quelle umzugehen haben. Der SolutionExplorer aktuell der einzige der die Diagramme vom Dateisystem und der PE anzeigen muss und dafür habe ich einfach zwei Services initialisiert. 

Die PE sollte auf jeden Fall früher oder später in Solutions denken, das würde auch im BPMN-Studio vieles erleichtern. Spätestens wenn es "interne" (Prozess erstellen & Prozess löschen) und "externe" Prozesse auf einer PE gibt würde sich dieses "umdenken" anbieten.

File: src/main.ts [hunk 2]
```diff
@@ -23,7 +28,15 @@ export function configure(aurelia: Aurelia): void {
     const fileInfo: IFileInfo = ipcRenderer.sendSync('get_opened_file');
     aurelia.container.registerInstance('FileContent', fileInfo);
     localStorage.setItem('processEngineRoute', `http://${newHost}`);
```
[the remark this answers is not in the document]
Comment (Vyperus) [reply src/main.ts]: Right now the "internal" PE inside the electron client is always connected via http. We can change this but this shouldn't be the scope of this PR. I Will add a comment here.

File: src/modules/diagram-detail/diagram-detail.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
+import {EventAggregator, Subscription} from 'aurelia-event-aggregator';
+import {inject} from 'aurelia-framework';
+import {Redirect, Router} from 'aurelia-router';
+
+import {IDiagram} from '@process-engine/solutionexplorer.contracts';
+import {ISolutionExplorerService} from '@process-engine/solutionexplorer.service.contracts';
+import * as download from 'downloadjs';
+import * as print from 'print-js';
+import * as beautify from 'xml-beautifier';
```
Comment (alexanderkasten) [inline src/modules/diagram-detail/diagram-detail.ts]: Remove unused imports

File: src/modules/diagram-detail/diagram-detail.ts [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+      }),
+      this._eventAggregator.subscribe(environment.events.diagramChange, () => {
+        this._diagramHasChanged = true;
+      }),
+    ];
+  }
+
+  public async canDeactivate(): Promise<Redirect> {
+
+    const _modal: Promise<boolean> = new Promise((resolve: Function, reject: Function): any => {
```
Comment (alexanderkasten) [inline src/modules/diagram-detail/diagram-detail.ts]: Is this `any` necessary ?

File: src/modules/diagram-detail/diagram-detail.ts [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+      subscription.dispose();
+    }
+
+    this._eventAggregator.publish(environment.events.navBar.hideTools);
+    this._eventAggregator.publish(environment.events.statusBar.hideXMLButton);
+  }
+
+  private async _saveDiagram(): Promise<void> {
+    try {
+    this.diagram.xml = await this.bpmnio.getXML();
```
Comment (alexanderkasten) [inline src/modules/diagram-detail/diagram-detail.ts]: Wrong indentation

File: src/modules/process-solution-panel/process-solution-panel.ts [hunk 6]
```diff
@@ -36,10 +64,12 @@ export class ProcessSolutionPanel {
       this._eventAggregator.subscribe(AuthenticationStateEvent.LOGOUT, () => {
         this._refreshProcesslist();
       }),
-      this._eventAggregator.subscribe(environment.events.refreshProcessDefs, () => {
-        this._refreshProcesslist();
-      }),
     ];
+
+    const solutionInputButton: HTMLElement = document.getElementById('solutionInputButton');
+    solutionInputButton.addEventListener('click', (event: any) => {
```
Comment (alexanderkasten) [inline src/modules/process-solution-panel/process-solution-panel.ts]: Use `KeyboardEvent` instead `any`

File: src/modules/process-solution-panel/process-solution-panel.ts [hunk 7]
```diff
@@ -51,7 +81,45 @@ export class ProcessSolutionPanel {
     window.localStorage.setItem('processSolutionExplorerHideState', 'hide');
   }
 
+  /**
+   * Handles the file input for the FileSystem Solutions.
+   * @param event A event that holds the files that were "uploaded" by the user.
+   * Currently there is no type for this kind of event.
+   */
+  public async onSolutionInputChange(event: any): Promise<void> {
```
Comment (alexanderkasten) [inline src/modules/process-solution-panel/process-solution-panel.ts]: Don't use any

File: src/modules/process-solution-panel/process-solution-panel.ts [hunk 8]
```diff
@@ -51,7 +81,45 @@ export class ProcessSolutionPanel {
...
+    await this._solutionExplorerServiceFileSystem.openSolution(event.target.files[0].path, this._identity);
+    const solution: ISolution = await this._solutionExplorerServiceFileSystem.loadSolution();
+    this.openedFileSystemSolutions.push(solution);
+    this.solutionInput.value = '';
+
```
Comment (alexanderkasten) [inline src/modules/process-solution-panel/process-solution-panel.ts]: Remove this empty line

File: src/modules/process-solution-panel/process-solution-panel.ts (hunk 7 above)
Comment (Vyperus) [reply src/modules/process-solution-panel/process-solution-panel.ts]: Sadly this parameter changes on the environment. Electron adds some properties to it so I wasn't able to find the correct Interface here.

File: src/modules/diagram-detail/diagram-detail.ts (hunk 4 above)
Comment (Vyperus) [reply src/modules/diagram-detail/diagram-detail.ts]: Sadly yes, the Promise will resolve with a boolean but not everytime. If you cancel the navigation it returns a Redirect.

Comment (Paulomart) [reply src/modules/diagram-detail/diagram-detail.ts]: Use `boolean | Redirect` as type

File: src/main.ts [hunk 9]
```diff
@@ -22,8 +26,20 @@ export function configure(aurelia: Aurelia): void {
...
     aurelia.container.registerInstance('FileContent', fileInfo);
+    /**
+     * Currently the internal PE is always connected via http.
+     * This will be subject to change.
+     */
     localStorage.setItem('processEngineRoute', `http://${newHost}`);
+
+    // Register SolutionExplorerFileSystemService
+    const fileSystemrepository: SolutionExplorerFileSystemRepository = new SolutionExplorerFileSystemRepository();
+    const filesystemSolutionexplorerService: SolutionExplorerService = new SolutionExplorerService(fileSystemrepository);
```
Comment (5ebastianMeier) [inline src/main.ts]: The case style on all of these is quite unique to what we usually use - why are things like `Solutionexplorer`, `filesystem` or `processengine` not written in proper Camel Case?

File: src/main.ts [hunk 10]
```diff
@@ -37,6 +53,17 @@ export function configure(aurelia: Aurelia): void {
...
+  /**
+   * Register Fake Identity for SolutionExplorer.
+   * TODO: Get real identity when IAM is finished.
+  */
+  const fakeIdentity: IIdentity = {
+    name: 'fakeIdentity',
+    id: 'fakeIdentity',
+    roles: [],
+  };
+  aurelia.container.registerInstance('Identity', fakeIdentity);
```
Comment (5ebastianMeier) [inline src/main.ts]: This should not be retrieved via injection, but via a service call.

If the identity changes, the injection results in obsolete state, that is likely not to get updated.
If retrieved through a service call, we can subscribe for updates as well.

File: src/modules/process-solution-panel/process-solution-panel.html [hunk 11]
```diff
@@ -1,9 +1,41 @@
...
+      </div>
+    </div>
+
+    <div class="solution-explorer__filesystem" show.bind="fileSystemIndexCardIsActive ">
+      <div class="process-explorer__heading">Solution Explorer
+        <button class="open-solution-button" id="solutionInputButton"><i class="fas fa-plus"></i></button>
+      </div>
+      <div repeat.for="solution of openedFileSystemSolutions">
+        <div class="solution-headline">
+          <i class="fa fa-folder-open" click.delegate="toggelVisibility($index)"></i>
```
Comment (5ebastianMeier) [inline src/modules/process-solution-panel/process-solution-panel.html]: typo *toggle

File: src/modules/process-solution-panel/process-solution-panel.ts [hunk 12]
```diff
@@ -1,32 +1,60 @@
...
-import {EventAggregator, Subscription} from 'aurelia-event-aggregator';
 import {AuthenticationStateEvent} from '../../contracts/index';
 import environment from '../../environment';
 
-@inject('BpmnStudioClient', EventAggregator)
+@inject(EventAggregator, Router, 'SolutionExplorerServiceProcessEngine', 'SolutionExplorerServiceFileSystem', 'Identity')
 export class ProcessSolutionPanel {
-  private _bpmnStudioClient: BpmnStudioClient;
+  public processes: IPagination<IProcessDefEntity>;
+  public processengineSolutionString: string;
```
Comment (5ebastianMeier) [inline src/modules/process-solution-panel/process-solution-panel.ts]: typo *processEngine

File: src/modules/process-solution-panel/process-solution-panel.ts [hunk 13]
```diff
@@ -36,10 +64,17 @@ export class ProcessSolutionPanel {
...
-        this._refreshProcesslist();
-      }),
     ];
+
+    const solutionInputButton: HTMLElement = document.getElementById('solutionInputButton');
+    solutionInputButton.addEventListener('click', () => {
+      this.solutionInput.click();
+    });
+
+    const firstSolutionInoutButton: HTMLElement = document.getElementById('openFirstSolutionButton');
```
Comment (5ebastianMeier) [inline src/modules/process-solution-panel/process-solution-panel.ts]: typo *Input

File: src/modules/process-solution-panel/process-solution-panel.ts [hunk 14]
```diff
@@ -51,7 +86,44 @@ export class ProcessSolutionPanel {
...
+  }
+
+  public async navigateToDiagramDetail(solution: ISolution, diagram: IDiagram): Promise<void> {
+    await this._solutionExplorerServiceFileSystem.openSolution(solution.uri, this._identity);
+    this._router.navigateToRoute('diagram-detail', {diagramName: diagram.name});
+  }
+
   private async _refreshProcesslist(): Promise<void> {
-    this.processes = await this._bpmnStudioClient.getProcessDefList();
+    this.processengineSolutionString = environment.bpmnStudioClient.baseRoute;
```
Comment (5ebastianMeier) [inline src/modules/process-solution-panel/process-solution-panel.ts]: typo *processEngine

**Pull request merged**